File: editorial/chap05/answer_B30.py
# a の b 乗を m で割った余りを返す関数
def Power(a, b, m):
	p = a
	Answer = 1
	for i in range(30):
		wari = 2 ** i
		if (b // wari) % 2 == 1:
			Answer = (Answer * p) % m # a の 2^i 乗が掛けられるとき
		p = (p * p) % m
	return Answer

# a÷b を m で割った余りを返す関数
def Division(a, b, m):
	return (a * Power(b, m - 2, m)) % m

# nCr mod 1000000007 を返す関数
def ncr(n, r):
	M = 1000000007;

	# 手順 1: 分子 a を求める
	a = 1;
	for i in range(1, n+1):
		a = (a*i) % M

	# 手順 2: 分母 b を求める
	b = 1;
	for i in range(1, r+1):
		b = (b * i) % M
	for i in range(1, n-r+1):
		b = (b * i) % M

	# 手順 3: 答えを求める
	return Division(a, b, M);

# 入力
H, W = map(int, input().split())

# 出力
print(ncr(H + W - 2, W - 1))

File: editorial/chap05/cpp/answer_B30.cpp
#include <iostream>
using namespace std;

// a の b 乗を m で割った余りを返す関数
// 変数 a は a^1 → a^2 → a^4 → a^8 → a^16 → ･･･ と変化
long long Power(long long a, long long b, long long m) {
	long long p = a, Answer = 1;
	for (int i = 0; i < 30; i++) {
		int wari = (1 << i);
		if ((b / wari) % 2 == 1) {
			Answer = (Answer * p) % m; // 「a の 2^i 乗」が掛けられるとき
		}
		p = (p * p) % m;
	}
	return Answer;
}

// a ÷ b を m で割った余りを返す関数
long long Division(long long a, long long b, long long m) {
	return (a * Power(b, m - 2, m)) % m;
}

// nCr mod 1000000007 を返す関数
long long ncr(int n, int r) {
	const long long M = 1000000007;

	// 手順 1: 分子 a を求める
	long long a = 1;
	for (int i = 1; i <= n; i++) a = (a * i) % M;

	// 手順 2: 分母 b を求める
	long long b = 1;
	for (int i = 1; i <= r; i++) b = (b * i) % M;
	for (int i = 1; i <= n - r; i++) b = (b * i) % M;

	// 手順 3: 答えを求める
	return Division(a, b, M);
}

int main() {
	// 入力
	long long H, W;
	cin >> H >> W;

	// 出力
	cout << ncr(H + W - 2, W - 1) << endl;
	return 0;
}
